A76-ring-hash-improvements.md

```diff
@@ -0,0 +1,169 @@
+
+At pick time:
+- If `request_metadata_key` is not empty, and the request metadata has a
+non-empty value, then the request hash key will be set to this value. If
+`request_metadata_key` contains multiple values, then values are joined with a
+comma `,` character between each value before hashing.
+- If `request_metadata_key` is not empty, and the request has no value or an
+empty value associated with the metadata key defined, then the picker will
+generate a random hash for it. The use of a random hash key will effectively
+sends the request to a random endpoint.
+- If `request_metadata_key` is empty, then the request hash key will be based on
+the xDS hash policy in RDS, which keeps the existing LB configuration for ring
+hash working as before with xDS.
+- If `request_metadata_key` is empty but the xDS configuration does not provide
+the hash key, then the picker will generate a random hash for it to select a
+random endpoint, which matches the current behavior for xDS.
+
+
+### Explicitly setting the endpoint hash key
+
+The `ring_hash` policy will be changed such that the hash key used for
+determining the locations of each endpoint on the ring will be extracted from a
+pre-defined endpoint attribute called `hash_key`. If this attribute is set, then
+the endpoint is placed on the ring by hashing its value. If this attribute is
+not set or empty, then the endpoint IP address is hashed, matching the current
+behavior. The locations of an existing endpoint on the ring is updated if its
+`hash_key` endpoint attribute changes.
+
+The cluster resolver LB policy responsible for translating EDS responses into
+resolver endpoints will be changed to set the `hash_key` endpoint attribute to
+the value of [LbEndpoint.Metadata][LbEndpoint.Metadata] `envoy.lb` `hash_key`
+field, as described in [Envoy's documentation for the ring hash load
+balancer][envoy-ring-hash].
+
+### LB Policy Config Changes
+
+After the addition of this field, the `ring_hash` LB policy config will be:
+
+```proto
+    message RingHashLoadBalancingConfig {
+      // A client-side cap will limit these values.  If either of these values
+      // are greater than the client-side cap, they will be treated as the
+      // client-side cap.  The default cap is 4096.
+      uint64 min_ring_size = 1;        // Optional, defaults to 1024.
+      uint64 max_ring_size = 2;        // Optional, defaults to 4096, max is 8M.
+
+      string request_metadata_key = 3; // Optional, unused if unset.
+    }
+```
+
+### Temporary environment variable protection
+
+Explicitly setting the request hash key cannot possibly lead to problem with
+existing deployment because the new behavior requires setting a load balancing
+policy configuration field that did not exist before. Therefore, it is not gated
+behind an environment variable.
+
+The second behavior change will be enabled by the
+`GRPC_EXPERIMENTAL_XDS_RING_HASH_ENDPOINT_HASH_KEY` environment variable. This
```

Comment:
I think this should be structured as more of a migration tool than an experiment-enabling feature.  I suggest the following text for this paragraph:

"""
Adding support for the `hash_key` in xDS endpoint metadata could potentially break existing clients whose control plane is setting this key, because upgrading the client to a new version of gRPC would automatically cause the key to start being used.  We expect that this change will not cause problems for most users, but just in case there is a problem, we will provide a migration story by supporting a temporary mechanism to tell gRPC to ignore the `hash_key` endpoint metadata.  This will be enabled by setting the `GRPC_XDS_HASH_KEY_BACKWARD_COMPAT` environment variable to true. This mechanism will be supported for a couple of gRPC releases but will be removed in the long run.
"""

Reply:
I took your wording here.